File: Data Structures & Algorithms/coin-change/submission-1.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        cache = {}
        def helper(n):
            nonlocal cache
            if n < 0:
                return float('inf')
            if n == 0:
                return 0
            
            if n in cache:
                return cache[n]

            x = float('inf')
            for i in coins:
                x = min(x, 1 + helper(n-i))
            print(n, x)
            cache[n] = x
            return x
        res = helper(amount)
        return res if res != float('inf') else -1

File: Data Structures & Algorithms/longest-palindromic-substring/submission-1.cpp
class Solution {
public:
    string helper(string s, int i, int j){
        if(i < 0 && j >= s.size()) return s;
        if(i < 0) return s.substr(0, j);
        if(j >= s.size()) return s.substr(i+1, s.size() - i);
        if(s[i] == s[j]){
            return helper(s, i-1, j+1);
        }
        return s.substr(i+1, j-i-1);
    }
    string longestPalindrome(string s) {
        string res;
        for(int i = 0; i < s.size(); i++){
            string temp = helper(s, i, i);
            if(temp.size() > res.size()) res = temp;
            temp = helper(s, i, i+1);
            if(temp.size() > res.size()) res = temp;
        }
        return res;
    }
};
